gobblin-core/src/main/java/gobblin/source/extractor/extract/sftp/SftpLightWeightFileSystem.java

```diff
@@ -0,0 +1,268 @@
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.source.extractor.extract.sftp;
+
+import gobblin.configuration.State;
+import gobblin.source.extractor.filebased.FileBasedHelperException;
+import gobblin.util.HadoopUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.List;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.BufferedFSInputStream;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FSInputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathFilter;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.util.Progressable;
+
+import com.google.common.collect.Lists;
+import com.jcraft.jsch.ChannelExec;
+import com.jcraft.jsch.SftpATTRS;
+import com.jcraft.jsch.SftpException;
+
+/**
+ * A {@link FileSystem} implementation that provides the {@link FileSystem} interface for an SFTP server.
+ * Uses {@link SftpFsHelper} internally to connect to the SFPT server.
+ */
+public class SftpLightWeightFileSystem extends FileSystem {
+
+  private static final URI NAME = URI.create("sftp:///");
+  private SftpFsHelper fsHelper;
+
+  private static final PathFilter VALID_PATH_FILTER = new PathFilter() {
+    @Override
+    public boolean accept(Path path) {
+      if (path == null) {
+        return false;
+      }
+      if (StringUtils.isBlank(path.toString())) {
+        return false;
+      }
+      if (path.toString().equals(".")) {
+        return false;
+      }
+      if (path.toString().equals("..")) {
+        return false;
+      }
+      return true;
+    }
+  };
+
+  @Override
+  public void initialize(URI name, Configuration conf)
+      throws IOException {
+    super.initialize(name, conf);
+    State state = HadoopUtils.getStateFromConf(conf);
+    fsHelper = new SftpFsHelper(state);
+    try {
+      fsHelper.connect();
+    } catch (FileBasedHelperException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public boolean delete(Path path) throws IOException {
+
+    try {
+      if (getFileStatus(path).isDir()) {
+        fsHelper.getSftpChannel().rmdir(path.toString());
+      } else {
+        fsHelper.getSftpChannel().rm(path.toString());
+      }
+    } catch (SftpException e) {
+      throw new IOException(e);
+    }
+    return true;
+  }
+
+  @Override
+  public boolean delete(Path path, boolean recursive) throws IOException {
+    if (!recursive) {
+      throw new UnsupportedOperationException("Non recursive delete is not supported on this file system");
```

Comment:
I think it is better to either completely implement is or throw UnsupportOperationException. Otherwise it confuses the caller as it works for "some" cases.